{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Flux Balance Analysis using COBRApy\n",
    "\n",
    "Simulations using flux balance analysis can be solved using COBRApy model.optimize() command. This module will maximize or minimize (maximizing is the default) flux based on the objective reactions.\n",
    "\n",
    "To start with, import the *E.coli* core model from the BIGG database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Academic license - for non-commercial use only - expires 2022-12-07\n",
      "Using license file C:\\Users\\scott\\gurobi.lic\n"
     ]
    }
   ],
   "source": [
    "import cobra.test\n",
    "from cobrapy_bigg_client import client\n",
    "model = client.download_model('e_coli_core', save=False) # Load model from BIGG database\n",
    "model.solver = 'glpk'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running a FBA simulation\n",
    "\n",
    "Place the results of the simulation in an object called \"solution\" and then print the result of the simulation which is the numerical value of the objective function which in the case corresponds to the growth rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Solution 0.874 at 0x129468075e0>\n"
     ]
    }
   ],
   "source": [
    "solution = model.optimize()\n",
    "print(solution)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The **model.optimize()** function will return a Solution object. A solution object has several attributes:\n",
    "\n",
    "- objective_value: the objective value\n",
    "- status: the status from the linear programming solver\n",
    "- fluxes: a pandas series with flux indexed by reaction identifier. The flux for a reaction variable is the difference of the primal values for the forward and reverse reaction variables.\n",
    "- shadow_prices: a pandas series with shadow price indexed by the metabolite identifier.\n",
    "\n",
    "For example, after the last call to model.optimize(), if the optimization succeeds it's status will be optimal. In case the model is infeasible an error is raised."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The objective value can also be found through either"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8739215069684303"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solution.objective_value"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "or"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8739215069684303"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.optimize().objective_value"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The solvers that can be used with COBRApy are so fast that for many small to mid-size models computing the solution can be even faster than it takes to collect the values from the solver and convert to them python objects. With model.optimize, COBRApy gathers values for all reactions and metabolites and that can take a significant amount of time if done repeatedly. If only the flux value of a single reaction or the objective, it is faster to use **model.slim_optimize()** which only does the optimization and returns the objective value leaving it up to you to fetch other values that you may need.\n",
    "\n",
    "This can be seen below in the following examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 2.01 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.8739215069684303"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "model.optimize().objective_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wall time: 0 ns\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.8739215069684303"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "model.slim_optimize()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyzing FBA solutions\n",
    "Models solved using FBA can be further analyzed by using summary methods, which produce printed text that gives a quick representation of optimized model behavior. Calling the summary method on the entire model produces information on the input and output behavior of the model, along with the optimized objective."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ecoli_core_w_GAM = 0.8739215069684297</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>100.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>4.765</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>21.8</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>3.215</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-22.81</td>\n",
       "      <td>1</td>\n",
       "      <td>100.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-29.18</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-17.53</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x129455242b0>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As can be seen, this summary produces the optimized objective plus all the flux values of all the exchange reactions that are actively uptaking flux into the cell as well as the flux values of the exhange reactions that are secreting flux into the  extracellular space."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Individual Reaction Flux Values\n",
    "\n",
    "The flux though a given reaction can be determined as follows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.7581774441067866"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.PYK.flux # This does not work with the Gurobi solver"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "or"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.7581774441068045"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solution.fluxes.PYK"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The reduced cost of a given reaction is given by"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-1.9999999999999991"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from cobra import Solution\n",
    "model.reactions.PYK.reduced_cost # This does not work with the gurobi solver"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Printing all the flux values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PFK                          7.477382\n",
       "PFL                          0.000000\n",
       "PGI                          4.860861\n",
       "PGK                        -16.023526\n",
       "PGL                          4.959985\n",
       "ACALD                        0.000000\n",
       "AKGt2r                       0.000000\n",
       "PGM                        -14.716140\n",
       "PIt2r                        3.214895\n",
       "ALCD2x                       0.000000\n",
       "ACALDt                       0.000000\n",
       "ACKr                         0.000000\n",
       "PPC                          2.504309\n",
       "ACONTa                       6.007250\n",
       "ACONTb                       6.007250\n",
       "ATPM                         8.390000\n",
       "PPCK                         0.000000\n",
       "ACt2r                        0.000000\n",
       "PPS                          0.000000\n",
       "ADK1                         0.000000\n",
       "AKGDH                        5.064376\n",
       "ATPS4r                      45.514010\n",
       "PTAr                         0.000000\n",
       "PYK                          1.758177\n",
       "BIOMASS_Ecoli_core_w_GAM     0.873922\n",
       "PYRt2                        0.000000\n",
       "CO2t                       -22.809833\n",
       "RPE                          2.678482\n",
       "CS                           6.007250\n",
       "RPI                         -2.281503\n",
       "SUCCt2_2                     0.000000\n",
       "CYTBD                       43.598985\n",
       "D_LACt2                      0.000000\n",
       "ENO                         14.716140\n",
       "SUCCt3                       0.000000\n",
       "ETOHt2r                      0.000000\n",
       "SUCDi                        5.064376\n",
       "SUCOAS                      -5.064376\n",
       "TALA                         1.496984\n",
       "THD2                         0.000000\n",
       "TKT1                         1.496984\n",
       "TKT2                         1.181498\n",
       "TPI                          7.477382\n",
       "EX_ac_e                      0.000000\n",
       "EX_acald_e                   0.000000\n",
       "EX_akg_e                     0.000000\n",
       "EX_co2_e                    22.809833\n",
       "EX_etoh_e                    0.000000\n",
       "EX_for_e                     0.000000\n",
       "EX_fru_e                     0.000000\n",
       "EX_fum_e                     0.000000\n",
       "EX_glc__D_e                -10.000000\n",
       "EX_gln__L_e                  0.000000\n",
       "EX_glu__L_e                  0.000000\n",
       "EX_h_e                      17.530865\n",
       "EX_h2o_e                    29.175827\n",
       "EX_lac__D_e                  0.000000\n",
       "EX_mal__L_e                  0.000000\n",
       "EX_nh4_e                    -4.765319\n",
       "EX_o2_e                    -21.799493\n",
       "EX_pi_e                     -3.214895\n",
       "EX_pyr_e                     0.000000\n",
       "EX_succ_e                    0.000000\n",
       "FBA                          7.477382\n",
       "FBP                          0.000000\n",
       "FORt2                        0.000000\n",
       "FORt                         0.000000\n",
       "FRD7                         0.000000\n",
       "FRUpts2                      0.000000\n",
       "FUM                          5.064376\n",
       "FUMt2_2                      0.000000\n",
       "G6PDH2r                      4.959985\n",
       "GAPD                        16.023526\n",
       "GLCpts                      10.000000\n",
       "GLNS                         0.223462\n",
       "GLNabc                       0.000000\n",
       "GLUDy                       -4.541857\n",
       "GLUN                         0.000000\n",
       "GLUSy                        0.000000\n",
       "GLUt2r                       0.000000\n",
       "GND                          4.959985\n",
       "H2Ot                       -29.175827\n",
       "ICDHyr                       6.007250\n",
       "ICL                          0.000000\n",
       "LDH_D                        0.000000\n",
       "MALS                         0.000000\n",
       "MALt2_2                      0.000000\n",
       "MDH                          5.064376\n",
       "ME1                          0.000000\n",
       "ME2                          0.000000\n",
       "NADH16                      38.534610\n",
       "NADTRHD                      0.000000\n",
       "NH4t                         4.765319\n",
       "O2t                         21.799493\n",
       "PDH                          9.282533\n",
       "Name: fluxes, dtype: float64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_rows', 500)\n",
    "solution.fluxes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Printing all the reduced costs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PFK                        -1.257675e-17\n",
       "PFL                         2.081668e-17\n",
       "PGI                         0.000000e+00\n",
       "PGK                         6.938894e-18\n",
       "PGL                         1.528725e-17\n",
       "ACALD                      -1.431147e-17\n",
       "AKGt2r                      8.673617e-18\n",
       "PGM                        -0.000000e+00\n",
       "PIt2r                      -7.806256e-18\n",
       "ALCD2x                     -2.125036e-17\n",
       "ACALDt                      0.000000e+00\n",
       "ACKr                        0.000000e+00\n",
       "PPC                         2.222614e-17\n",
       "ACONTa                      1.376937e-17\n",
       "ACONTb                      1.084202e-19\n",
       "ATPM                       -1.018497e-02\n",
       "PPCK                       -1.018497e-02\n",
       "ACt2r                      -1.734723e-18\n",
       "PPS                        -1.018497e-02\n",
       "ADK1                       -0.000000e+00\n",
       "AKGDH                      -1.734723e-18\n",
       "ATPS4r                      1.192622e-18\n",
       "PTAr                        1.647987e-17\n",
       "PYK                         5.637851e-18\n",
       "BIOMASS_Ecoli_core_w_GAM    4.857226e-17\n",
       "PYRt2                       1.561251e-17\n",
       "CO2t                       -0.000000e+00\n",
       "RPE                        -4.996004e-16\n",
       "CS                          1.409463e-18\n",
       "RPI                        -0.000000e+00\n",
       "SUCCt2_2                    2.428613e-17\n",
       "CYTBD                      -1.409463e-18\n",
       "D_LACt2                     1.734723e-18\n",
       "ENO                        -1.084202e-19\n",
       "SUCCt3                     -7.638729e-03\n",
       "ETOHt2r                     1.734723e-18\n",
       "SUCDi                      -2.168404e-18\n",
       "SUCOAS                     -2.602085e-18\n",
       "TALA                        2.220446e-16\n",
       "THD2                       -2.546243e-03\n",
       "TKT1                        8.049117e-16\n",
       "TKT2                       -6.106227e-16\n",
       "TPI                         0.000000e+00\n",
       "EX_ac_e                    -4.583237e-02\n",
       "EX_acald_e                 -6.874856e-02\n",
       "EX_akg_e                   -1.222197e-01\n",
       "EX_co2_e                    0.000000e+00\n",
       "EX_etoh_e                  -7.893353e-02\n",
       "EX_for_e                   -1.527746e-02\n",
       "EX_fru_e                   -1.833295e-01\n",
       "EX_fum_e                   -9.166475e-02\n",
       "EX_glc__D_e                -1.833295e-01\n",
       "EX_gln__L_e                -1.400434e-01\n",
       "EX_glu__L_e                -1.374971e-01\n",
       "EX_h_e                      4.336809e-19\n",
       "EX_h2o_e                    1.084202e-19\n",
       "EX_lac__D_e                -8.147977e-02\n",
       "EX_mal__L_e                -9.166475e-02\n",
       "EX_nh4_e                   -1.387779e-17\n",
       "EX_o2_e                     0.000000e+00\n",
       "EX_pi_e                    -4.163336e-17\n",
       "EX_pyr_e                   -6.874856e-02\n",
       "EX_succ_e                  -9.675723e-02\n",
       "FBA                         0.000000e+00\n",
       "FBP                        -1.018497e-02\n",
       "FORt2                      -2.546243e-03\n",
       "FORt                       -0.000000e+00\n",
       "FRD7                        4.336809e-18\n",
       "FRUpts2                     6.938894e-18\n",
       "FUM                         1.084202e-19\n",
       "FUMt2_2                     3.122502e-17\n",
       "G6PDH2r                     9.844556e-17\n",
       "GAPD                       -4.336809e-19\n",
       "GLCpts                      0.000000e+00\n",
       "GLNS                       -2.914335e-16\n",
       "GLNabc                     -6.938894e-17\n",
       "GLUDy                       4.163336e-17\n",
       "GLUN                       -1.018497e-02\n",
       "GLUSy                      -1.018497e-02\n",
       "GLUt2r                     -2.775558e-17\n",
       "GND                         9.714451e-17\n",
       "H2Ot                       -0.000000e+00\n",
       "ICDHyr                      2.775558e-17\n",
       "ICL                        -2.546243e-03\n",
       "LDH_D                      -1.951564e-17\n",
       "MALS                       -1.387779e-17\n",
       "MALt2_2                     3.122502e-17\n",
       "MDH                        -1.257675e-17\n",
       "ME1                        -1.018497e-02\n",
       "ME2                        -7.638729e-03\n",
       "NADH16                      0.000000e+00\n",
       "NADTRHD                    -2.546243e-03\n",
       "NH4t                       -1.387779e-17\n",
       "O2t                         0.000000e+00\n",
       "PDH                        -1.561251e-17\n",
       "Name: reduced_costs, dtype: float64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solution.reduced_costs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating a table that includes the reaction IDs, reaction names, reaction flux, and reduced costs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction ID</th>\n",
       "      <th>Reaction Name</th>\n",
       "      <th>Reaction Flux</th>\n",
       "      <th>Reaction Reduced Cost</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PFK</td>\n",
       "      <td>Phosphofructokinase</td>\n",
       "      <td>7.477382e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PFL</td>\n",
       "      <td>Pyruvate formate lyase</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGI</td>\n",
       "      <td>Glucose-6-phosphate isomerase</td>\n",
       "      <td>4.860861e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGK</td>\n",
       "      <td>Phosphoglycerate kinase</td>\n",
       "      <td>-1.602353e+01</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PGL</td>\n",
       "      <td>6-phosphogluconolactonase</td>\n",
       "      <td>4.959985e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACALD</td>\n",
       "      <td>Acetaldehyde dehydrogenase (acetylating)</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AKGt2r</td>\n",
       "      <td>2 oxoglutarate reversible transport via symport</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PGM</td>\n",
       "      <td>Phosphoglycerate mutase</td>\n",
       "      <td>-1.471614e+01</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PIt2r</td>\n",
       "      <td>Phosphate reversible transport via symport</td>\n",
       "      <td>3.214895e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ALCD2x</td>\n",
       "      <td>Alcohol dehydrogenase (ethanol)</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ACALDt</td>\n",
       "      <td>Acetaldehyde reversible transport</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ACKr</td>\n",
       "      <td>Acetate kinase</td>\n",
       "      <td>-1.937884e-14</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PPC</td>\n",
       "      <td>Phosphoenolpyruvate carboxylase</td>\n",
       "      <td>2.504309e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ACONTa</td>\n",
       "      <td>Aconitase (half-reaction A, Citrate hydro-lyase)</td>\n",
       "      <td>6.007250e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ACONTb</td>\n",
       "      <td>Aconitase (half-reaction B, Isocitrate hydro-l...</td>\n",
       "      <td>6.007250e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ATPM</td>\n",
       "      <td>ATP maintenance requirement</td>\n",
       "      <td>8.390000e+00</td>\n",
       "      <td>12.488889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>PPCK</td>\n",
       "      <td>Phosphoenolpyruvate carboxykinase</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>14.488889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ACt2r</td>\n",
       "      <td>Acetate reversible transport via proton symport</td>\n",
       "      <td>-1.937884e-14</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PPS</td>\n",
       "      <td>Phosphoenolpyruvate synthase</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>12.488889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ADK1</td>\n",
       "      <td>Adenylate kinase</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AKGDH</td>\n",
       "      <td>2-Oxogluterate dehydrogenase</td>\n",
       "      <td>5.064376e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ATPS4r</td>\n",
       "      <td>ATP synthase (four protons for one ATP)</td>\n",
       "      <td>4.551401e+01</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PTAr</td>\n",
       "      <td>Phosphotransacetylase</td>\n",
       "      <td>1.937884e-14</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PYK</td>\n",
       "      <td>Pyruvate kinase</td>\n",
       "      <td>1.758177e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>BIOMASS_Ecoli_core_w_GAM</td>\n",
       "      <td>Biomass Objective Function with GAM</td>\n",
       "      <td>8.739215e-01</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>PYRt2</td>\n",
       "      <td>Pyruvate transport in via proton symport</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CO2t</td>\n",
       "      <td>CO2 transporter via diffusion</td>\n",
       "      <td>-2.280983e+01</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>RPE</td>\n",
       "      <td>Ribulose 5-phosphate 3-epimerase</td>\n",
       "      <td>2.678482e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CS</td>\n",
       "      <td>Citrate synthase</td>\n",
       "      <td>6.007250e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>RPI</td>\n",
       "      <td>Ribose-5-phosphate isomerase</td>\n",
       "      <td>-2.281503e+00</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SUCCt2_2</td>\n",
       "      <td>Succinate transport via proton symport (2 H)</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CYTBD</td>\n",
       "      <td>Cytochrome oxidase bd (ubiquinol-8: 2 protons)</td>\n",
       "      <td>4.359899e+01</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>D_LACt2</td>\n",
       "      <td>D lactate transport via proton symport</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENO</td>\n",
       "      <td>Enolase</td>\n",
       "      <td>1.471614e+01</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SUCCt3</td>\n",
       "      <td>Succinate transport out via proton antiport</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>9.866667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ETOHt2r</td>\n",
       "      <td>Ethanol reversible transport via proton symport</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>SUCDi</td>\n",
       "      <td>Succinate dehydrogenase (irreversible)</td>\n",
       "      <td>5.064376e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>SUCOAS</td>\n",
       "      <td>Succinyl-CoA synthetase (ADP-forming)</td>\n",
       "      <td>-5.064376e+00</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>TALA</td>\n",
       "      <td>Transaldolase</td>\n",
       "      <td>1.496984e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>THD2</td>\n",
       "      <td>NAD(P) transhydrogenase</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.822222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>TKT1</td>\n",
       "      <td>Transketolase</td>\n",
       "      <td>1.496984e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>TKT2</td>\n",
       "      <td>Transketolase</td>\n",
       "      <td>1.181498e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>TPI</td>\n",
       "      <td>Triose-phosphate isomerase</td>\n",
       "      <td>7.477382e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>EX_ac_e</td>\n",
       "      <td>Acetate exchange</td>\n",
       "      <td>1.937884e-14</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>EX_acald_e</td>\n",
       "      <td>Acetaldehyde exchange</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>EX_akg_e</td>\n",
       "      <td>2-Oxoglutarate exchange</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>26.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>CO2 exchange</td>\n",
       "      <td>2.280983e+01</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>EX_etoh_e</td>\n",
       "      <td>Ethanol exchange</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.088889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>EX_for_e</td>\n",
       "      <td>Formate exchange</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.733333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>EX_fru_e</td>\n",
       "      <td>D-Fructose exchange</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>12.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>EX_fum_e</td>\n",
       "      <td>Fumarate exchange</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>17.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>D-Glucose exchange</td>\n",
       "      <td>-1.000000e+01</td>\n",
       "      <td>10.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>EX_gln__L_e</td>\n",
       "      <td>L-Glutamine exchange</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>53.822222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>EX_glu__L_e</td>\n",
       "      <td>L-Glutamate exchange</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>41.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>H+ exchange</td>\n",
       "      <td>1.753087e+01</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>H2O exchange</td>\n",
       "      <td>2.917583e+01</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>EX_lac__D_e</td>\n",
       "      <td>D-lactate exchange</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.711111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>EX_mal__L_e</td>\n",
       "      <td>L-Malate exchange</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>15.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>Ammonia exchange</td>\n",
       "      <td>-4.765319e+00</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>O2 exchange</td>\n",
       "      <td>-2.179949e+01</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>Phosphate exchange</td>\n",
       "      <td>-3.214895e+00</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>EX_pyr_e</td>\n",
       "      <td>Pyruvate exchange</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>EX_succ_e</td>\n",
       "      <td>Succinate exchange</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>12.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>FBA</td>\n",
       "      <td>Fructose-bisphosphate aldolase</td>\n",
       "      <td>7.477382e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>FBP</td>\n",
       "      <td>Fructose-bisphosphatase</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>14.488889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>FORt2</td>\n",
       "      <td>Formate transport in via proton symport</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.622222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>FORt</td>\n",
       "      <td>Formate transport via diffusion</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>FRD7</td>\n",
       "      <td>Fumarate reductase</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>FRUpts2</td>\n",
       "      <td>Fructose transport via PEP:Pyr PTS (f6p genera...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>FUM</td>\n",
       "      <td>Fumarase</td>\n",
       "      <td>5.064376e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>FUMt2_2</td>\n",
       "      <td>Fumarate transport via proton symport (2 H)</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>G6PDH2r</td>\n",
       "      <td>Glucose 6-phosphate dehydrogenase</td>\n",
       "      <td>4.959985e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>GAPD</td>\n",
       "      <td>Glyceraldehyde-3-phosphate dehydrogenase</td>\n",
       "      <td>1.602353e+01</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>GLCpts</td>\n",
       "      <td>D-glucose transport via PEP:Pyr PTS</td>\n",
       "      <td>1.000000e+01</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>GLNS</td>\n",
       "      <td>Glutamine synthetase</td>\n",
       "      <td>2.234617e-01</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>GLNabc</td>\n",
       "      <td>L-glutamine transport via ABC system</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>GLUDy</td>\n",
       "      <td>Glutamate dehydrogenase (NADP)</td>\n",
       "      <td>-4.541857e+00</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>GLUN</td>\n",
       "      <td>Glutaminase</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>14.488889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>GLUSy</td>\n",
       "      <td>Glutamate synthase (NADPH)</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>12.488889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>GLUt2r</td>\n",
       "      <td>L glutamate transport via proton symport  reve...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>GND</td>\n",
       "      <td>Phosphogluconate dehydrogenase</td>\n",
       "      <td>4.959985e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>H2Ot</td>\n",
       "      <td>H2O transport via diffusion</td>\n",
       "      <td>-2.917583e+01</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>ICDHyr</td>\n",
       "      <td>Isocitrate dehydrogenase (NADP)</td>\n",
       "      <td>6.007250e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>ICL</td>\n",
       "      <td>Isocitrate lyase</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.422222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>LDH_D</td>\n",
       "      <td>D-lactate dehydrogenase</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>MALS</td>\n",
       "      <td>Malate synthase</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>MALt2_2</td>\n",
       "      <td>Malate transport via proton symport (2 H)</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>MDH</td>\n",
       "      <td>Malate dehydrogenase</td>\n",
       "      <td>5.064376e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>ME1</td>\n",
       "      <td>Malic enzyme (NAD)</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>10.488889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>ME2</td>\n",
       "      <td>Malic enzyme (NADP)</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>9.066667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>NADH16</td>\n",
       "      <td>NADH dehydrogenase (ubiquinone-8 &amp; 3 protons)</td>\n",
       "      <td>3.853461e+01</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>NADTRHD</td>\n",
       "      <td>NAD transhydrogenase</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.422222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>NH4t</td>\n",
       "      <td>Ammonia reversible transport</td>\n",
       "      <td>4.765319e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>O2t</td>\n",
       "      <td>O2 transport  diffusion</td>\n",
       "      <td>2.179949e+01</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>PDH</td>\n",
       "      <td>Pyruvate dehydrogenase</td>\n",
       "      <td>9.282533e+00</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Reaction ID  \\\n",
       "0                        PFK   \n",
       "1                        PFL   \n",
       "2                        PGI   \n",
       "3                        PGK   \n",
       "4                        PGL   \n",
       "5                      ACALD   \n",
       "6                     AKGt2r   \n",
       "7                        PGM   \n",
       "8                      PIt2r   \n",
       "9                     ALCD2x   \n",
       "10                    ACALDt   \n",
       "11                      ACKr   \n",
       "12                       PPC   \n",
       "13                    ACONTa   \n",
       "14                    ACONTb   \n",
       "15                      ATPM   \n",
       "16                      PPCK   \n",
       "17                     ACt2r   \n",
       "18                       PPS   \n",
       "19                      ADK1   \n",
       "20                     AKGDH   \n",
       "21                    ATPS4r   \n",
       "22                      PTAr   \n",
       "23                       PYK   \n",
       "24  BIOMASS_Ecoli_core_w_GAM   \n",
       "25                     PYRt2   \n",
       "26                      CO2t   \n",
       "27                       RPE   \n",
       "28                        CS   \n",
       "29                       RPI   \n",
       "30                  SUCCt2_2   \n",
       "31                     CYTBD   \n",
       "32                   D_LACt2   \n",
       "33                       ENO   \n",
       "34                    SUCCt3   \n",
       "35                   ETOHt2r   \n",
       "36                     SUCDi   \n",
       "37                    SUCOAS   \n",
       "38                      TALA   \n",
       "39                      THD2   \n",
       "40                      TKT1   \n",
       "41                      TKT2   \n",
       "42                       TPI   \n",
       "43                   EX_ac_e   \n",
       "44                EX_acald_e   \n",
       "45                  EX_akg_e   \n",
       "46                  EX_co2_e   \n",
       "47                 EX_etoh_e   \n",
       "48                  EX_for_e   \n",
       "49                  EX_fru_e   \n",
       "50                  EX_fum_e   \n",
       "51               EX_glc__D_e   \n",
       "52               EX_gln__L_e   \n",
       "53               EX_glu__L_e   \n",
       "54                    EX_h_e   \n",
       "55                  EX_h2o_e   \n",
       "56               EX_lac__D_e   \n",
       "57               EX_mal__L_e   \n",
       "58                  EX_nh4_e   \n",
       "59                   EX_o2_e   \n",
       "60                   EX_pi_e   \n",
       "61                  EX_pyr_e   \n",
       "62                 EX_succ_e   \n",
       "63                       FBA   \n",
       "64                       FBP   \n",
       "65                     FORt2   \n",
       "66                      FORt   \n",
       "67                      FRD7   \n",
       "68                   FRUpts2   \n",
       "69                       FUM   \n",
       "70                   FUMt2_2   \n",
       "71                   G6PDH2r   \n",
       "72                      GAPD   \n",
       "73                    GLCpts   \n",
       "74                      GLNS   \n",
       "75                    GLNabc   \n",
       "76                     GLUDy   \n",
       "77                      GLUN   \n",
       "78                     GLUSy   \n",
       "79                    GLUt2r   \n",
       "80                       GND   \n",
       "81                      H2Ot   \n",
       "82                    ICDHyr   \n",
       "83                       ICL   \n",
       "84                     LDH_D   \n",
       "85                      MALS   \n",
       "86                   MALt2_2   \n",
       "87                       MDH   \n",
       "88                       ME1   \n",
       "89                       ME2   \n",
       "90                    NADH16   \n",
       "91                   NADTRHD   \n",
       "92                      NH4t   \n",
       "93                       O2t   \n",
       "94                       PDH   \n",
       "\n",
       "                                        Reaction Name  Reaction Flux  \\\n",
       "0                                 Phosphofructokinase   7.477382e+00   \n",
       "1                              Pyruvate formate lyase   0.000000e+00   \n",
       "2                       Glucose-6-phosphate isomerase   4.860861e+00   \n",
       "3                             Phosphoglycerate kinase  -1.602353e+01   \n",
       "4                           6-phosphogluconolactonase   4.959985e+00   \n",
       "5            Acetaldehyde dehydrogenase (acetylating)   0.000000e+00   \n",
       "6     2 oxoglutarate reversible transport via symport   0.000000e+00   \n",
       "7                             Phosphoglycerate mutase  -1.471614e+01   \n",
       "8          Phosphate reversible transport via symport   3.214895e+00   \n",
       "9                     Alcohol dehydrogenase (ethanol)   0.000000e+00   \n",
       "10                  Acetaldehyde reversible transport   0.000000e+00   \n",
       "11                                     Acetate kinase  -1.937884e-14   \n",
       "12                    Phosphoenolpyruvate carboxylase   2.504309e+00   \n",
       "13   Aconitase (half-reaction A, Citrate hydro-lyase)   6.007250e+00   \n",
       "14  Aconitase (half-reaction B, Isocitrate hydro-l...   6.007250e+00   \n",
       "15                        ATP maintenance requirement   8.390000e+00   \n",
       "16                  Phosphoenolpyruvate carboxykinase   0.000000e+00   \n",
       "17    Acetate reversible transport via proton symport  -1.937884e-14   \n",
       "18                       Phosphoenolpyruvate synthase   0.000000e+00   \n",
       "19                                   Adenylate kinase   0.000000e+00   \n",
       "20                       2-Oxogluterate dehydrogenase   5.064376e+00   \n",
       "21            ATP synthase (four protons for one ATP)   4.551401e+01   \n",
       "22                              Phosphotransacetylase   1.937884e-14   \n",
       "23                                    Pyruvate kinase   1.758177e+00   \n",
       "24                Biomass Objective Function with GAM   8.739215e-01   \n",
       "25           Pyruvate transport in via proton symport   0.000000e+00   \n",
       "26                      CO2 transporter via diffusion  -2.280983e+01   \n",
       "27                   Ribulose 5-phosphate 3-epimerase   2.678482e+00   \n",
       "28                                   Citrate synthase   6.007250e+00   \n",
       "29                       Ribose-5-phosphate isomerase  -2.281503e+00   \n",
       "30       Succinate transport via proton symport (2 H)   0.000000e+00   \n",
       "31     Cytochrome oxidase bd (ubiquinol-8: 2 protons)   4.359899e+01   \n",
       "32             D lactate transport via proton symport   0.000000e+00   \n",
       "33                                            Enolase   1.471614e+01   \n",
       "34        Succinate transport out via proton antiport   0.000000e+00   \n",
       "35    Ethanol reversible transport via proton symport   0.000000e+00   \n",
       "36             Succinate dehydrogenase (irreversible)   5.064376e+00   \n",
       "37              Succinyl-CoA synthetase (ADP-forming)  -5.064376e+00   \n",
       "38                                      Transaldolase   1.496984e+00   \n",
       "39                            NAD(P) transhydrogenase   0.000000e+00   \n",
       "40                                      Transketolase   1.496984e+00   \n",
       "41                                      Transketolase   1.181498e+00   \n",
       "42                         Triose-phosphate isomerase   7.477382e+00   \n",
       "43                                   Acetate exchange   1.937884e-14   \n",
       "44                              Acetaldehyde exchange   0.000000e+00   \n",
       "45                            2-Oxoglutarate exchange   0.000000e+00   \n",
       "46                                       CO2 exchange   2.280983e+01   \n",
       "47                                   Ethanol exchange   0.000000e+00   \n",
       "48                                   Formate exchange   0.000000e+00   \n",
       "49                                D-Fructose exchange   0.000000e+00   \n",
       "50                                  Fumarate exchange   0.000000e+00   \n",
       "51                                 D-Glucose exchange  -1.000000e+01   \n",
       "52                               L-Glutamine exchange   0.000000e+00   \n",
       "53                               L-Glutamate exchange   0.000000e+00   \n",
       "54                                        H+ exchange   1.753087e+01   \n",
       "55                                       H2O exchange   2.917583e+01   \n",
       "56                                 D-lactate exchange   0.000000e+00   \n",
       "57                                  L-Malate exchange   0.000000e+00   \n",
       "58                                   Ammonia exchange  -4.765319e+00   \n",
       "59                                        O2 exchange  -2.179949e+01   \n",
       "60                                 Phosphate exchange  -3.214895e+00   \n",
       "61                                  Pyruvate exchange   0.000000e+00   \n",
       "62                                 Succinate exchange   0.000000e+00   \n",
       "63                     Fructose-bisphosphate aldolase   7.477382e+00   \n",
       "64                            Fructose-bisphosphatase   0.000000e+00   \n",
       "65            Formate transport in via proton symport   0.000000e+00   \n",
       "66                    Formate transport via diffusion   0.000000e+00   \n",
       "67                                 Fumarate reductase   0.000000e+00   \n",
       "68  Fructose transport via PEP:Pyr PTS (f6p genera...   0.000000e+00   \n",
       "69                                           Fumarase   5.064376e+00   \n",
       "70        Fumarate transport via proton symport (2 H)   0.000000e+00   \n",
       "71                  Glucose 6-phosphate dehydrogenase   4.959985e+00   \n",
       "72           Glyceraldehyde-3-phosphate dehydrogenase   1.602353e+01   \n",
       "73                D-glucose transport via PEP:Pyr PTS   1.000000e+01   \n",
       "74                               Glutamine synthetase   2.234617e-01   \n",
       "75               L-glutamine transport via ABC system   0.000000e+00   \n",
       "76                     Glutamate dehydrogenase (NADP)  -4.541857e+00   \n",
       "77                                        Glutaminase   0.000000e+00   \n",
       "78                         Glutamate synthase (NADPH)   0.000000e+00   \n",
       "79  L glutamate transport via proton symport  reve...   0.000000e+00   \n",
       "80                     Phosphogluconate dehydrogenase   4.959985e+00   \n",
       "81                        H2O transport via diffusion  -2.917583e+01   \n",
       "82                    Isocitrate dehydrogenase (NADP)   6.007250e+00   \n",
       "83                                   Isocitrate lyase   0.000000e+00   \n",
       "84                            D-lactate dehydrogenase   0.000000e+00   \n",
       "85                                    Malate synthase   0.000000e+00   \n",
       "86          Malate transport via proton symport (2 H)   0.000000e+00   \n",
       "87                               Malate dehydrogenase   5.064376e+00   \n",
       "88                                 Malic enzyme (NAD)   0.000000e+00   \n",
       "89                                Malic enzyme (NADP)   0.000000e+00   \n",
       "90      NADH dehydrogenase (ubiquinone-8 & 3 protons)   3.853461e+01   \n",
       "91                               NAD transhydrogenase   0.000000e+00   \n",
       "92                       Ammonia reversible transport   4.765319e+00   \n",
       "93                           O2 transport  diffusion    2.179949e+01   \n",
       "94                             Pyruvate dehydrogenase   9.282533e+00   \n",
       "\n",
       "    Reaction Reduced Cost  \n",
       "0               -2.000000  \n",
       "1               -2.000000  \n",
       "2               -2.000000  \n",
       "3                2.000000  \n",
       "4               -2.000000  \n",
       "5               -2.000000  \n",
       "6               -2.000000  \n",
       "7                2.000000  \n",
       "8               -2.000000  \n",
       "9               -2.000000  \n",
       "10              -2.000000  \n",
       "11               2.000000  \n",
       "12              -2.000000  \n",
       "13              -2.000000  \n",
       "14              -2.000000  \n",
       "15              12.488889  \n",
       "16              14.488889  \n",
       "17               2.000000  \n",
       "18              12.488889  \n",
       "19               2.000000  \n",
       "20              -2.000000  \n",
       "21              -2.000000  \n",
       "22              -2.000000  \n",
       "23              -2.000000  \n",
       "24              -2.000000  \n",
       "25              -2.000000  \n",
       "26               2.000000  \n",
       "27              -2.000000  \n",
       "28              -2.000000  \n",
       "29               2.000000  \n",
       "30              -2.000000  \n",
       "31              -2.000000  \n",
       "32              -2.000000  \n",
       "33              -2.000000  \n",
       "34               9.866667  \n",
       "35              -2.000000  \n",
       "36              -2.000000  \n",
       "37               2.000000  \n",
       "38              -2.000000  \n",
       "39               3.822222  \n",
       "40              -2.000000  \n",
       "41              -2.000000  \n",
       "42              -2.000000  \n",
       "43              -2.000000  \n",
       "44               5.600000  \n",
       "45              26.666667  \n",
       "46              -2.000000  \n",
       "47               4.088889  \n",
       "48               5.733333  \n",
       "49              12.400000  \n",
       "50              17.200000  \n",
       "51              10.400000  \n",
       "52              53.822222  \n",
       "53              41.200000  \n",
       "54              -2.000000  \n",
       "55              -2.000000  \n",
       "56               4.711111  \n",
       "57              15.200000  \n",
       "58               2.000000  \n",
       "59               2.000000  \n",
       "60               2.000000  \n",
       "61               3.600000  \n",
       "62              12.444444  \n",
       "63              -2.000000  \n",
       "64              14.488889  \n",
       "65               4.622222  \n",
       "66               2.000000  \n",
       "67               2.000000  \n",
       "68              -2.000000  \n",
       "69              -2.000000  \n",
       "70              -2.000000  \n",
       "71              -2.000000  \n",
       "72              -2.000000  \n",
       "73              -2.000000  \n",
       "74              -2.000000  \n",
       "75              -2.000000  \n",
       "76               2.000000  \n",
       "77              14.488889  \n",
       "78              12.488889  \n",
       "79              -2.000000  \n",
       "80              -2.000000  \n",
       "81               2.000000  \n",
       "82              -2.000000  \n",
       "83               1.422222  \n",
       "84              -2.000000  \n",
       "85              -2.000000  \n",
       "86              -2.000000  \n",
       "87              -2.000000  \n",
       "88              10.488889  \n",
       "89               9.066667  \n",
       "90              -2.000000  \n",
       "91               1.422222  \n",
       "92              -2.000000  \n",
       "93              -2.000000  \n",
       "94              -2.000000  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "reaction_names = [r.name for r in model.reactions]\n",
    "reaction_ids = [r.id for r in model.reactions]\n",
    "reaction_flux = [r.flux for r in model.reactions]\n",
    "reaction_RC = [r.reduced_cost for r in model.reactions]\n",
    "reactionList = {'Reaction ID': reaction_ids,\n",
    "                'Reaction Name': reaction_names,\n",
    "                'Reaction Flux': reaction_flux,\n",
    "                'Reaction Reduced Cost': reaction_RC,\n",
    "               }\n",
    "\n",
    "df = pd.DataFrame(reactionList, columns= ['Reaction ID','Reaction Name','Reaction Flux','Reaction Reduced Cost'])\n",
    "pd.set_option('display.max_rows', 500)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Individual Metabolite Flux Values\n",
    "\n",
    "In addition, the input-output behavior of individual metabolites can also be inspected using summary methods. For example, the following commands can be used to examine the flux associated with the reactions that are both producing and consuming the given metabolite."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>nadh_c</h3><p>C21H27N7O14P2</p><h4>Producing Reactions</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Percent</th>\n",
       "      <th>Flux</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>13.14%</td>\n",
       "      <td>5.064</td>\n",
       "      <td>AKGDH</td>\n",
       "      <td>akg_c + coa_c + nad_c --&gt; co2_c + nadh_c + succoa_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8.04%</td>\n",
       "      <td>3.1</td>\n",
       "      <td>BIOMASS_Ecoli_core_w_GAM</td>\n",
       "      <td>1.496 3pg_c + 3.7478 accoa_c + 59.81 atp_c + 0.361 e4p_c + 0.0709 f6p_c + 0.129 g3p_c + 0.205 g6p_c + 0.2557 gln__L_c + 4.9414 glu__L_c + 59.81 h2o_c + 3.547 nad_c + 13.0279 nadph_c + 1.7867 oaa_c + 0.5191 pep_c + 2.8328 pyr_c + 0.8977 r5p_c --&gt; 59.81 adp_c + 4.1182 akg_c + 3.7478 coa_c + 59.81 h_c + 3.547 nadh_c + 13.0279 nadp_c + 59.81 pi_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41.58%</td>\n",
       "      <td>16.02</td>\n",
       "      <td>GAPD</td>\n",
       "      <td>g3p_c + nad_c + pi_c &lt;=&gt; 13dpg_c + h_c + nadh_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13.14%</td>\n",
       "      <td>5.064</td>\n",
       "      <td>MDH</td>\n",
       "      <td>mal__L_c + nad_c &lt;=&gt; h_c + nadh_c + oaa_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24.09%</td>\n",
       "      <td>9.283</td>\n",
       "      <td>PDH</td>\n",
       "      <td>coa_c + nad_c + pyr_c --&gt; accoa_c + co2_c + nadh_c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Consuming Reactions</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Percent</th>\n",
       "      <th>Flux</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>100.00%</td>\n",
       "      <td>-38.53</td>\n",
       "      <td>NADH16</td>\n",
       "      <td>4.0 h_c + nadh_c + q8_c --&gt; 3.0 h_e + nad_c + q8h2_c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.metabolite_summary.MetaboliteSummary at 0x12946836bb0>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.nadh_c.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will be helpful in determining the main energy production and consumption reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>atp_c</h3><p>C10H12N5O13P3</p><h4>Producing Reactions</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Percent</th>\n",
       "      <th>Flux</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>66.58%</td>\n",
       "      <td>45.51</td>\n",
       "      <td>ATPS4r</td>\n",
       "      <td>adp_c + 4.0 h_e + pi_c &lt;=&gt; atp_c + h2o_c + 3.0 h_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>23.44%</td>\n",
       "      <td>16.02</td>\n",
       "      <td>PGK</td>\n",
       "      <td>3pg_c + atp_c &lt;=&gt; 13dpg_c + adp_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2.57%</td>\n",
       "      <td>1.758</td>\n",
       "      <td>PYK</td>\n",
       "      <td>adp_c + h_c + pep_c --&gt; atp_c + pyr_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7.41%</td>\n",
       "      <td>5.064</td>\n",
       "      <td>SUCOAS</td>\n",
       "      <td>atp_c + coa_c + succ_c &lt;=&gt; adp_c + pi_c + succoa_c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Consuming Reactions</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Percent</th>\n",
       "      <th>Flux</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>12.27%</td>\n",
       "      <td>-8.39</td>\n",
       "      <td>ATPM</td>\n",
       "      <td>atp_c + h2o_c --&gt; adp_c + h_c + pi_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>76.46%</td>\n",
       "      <td>-52.27</td>\n",
       "      <td>BIOMASS_Ecoli_core_w_GAM</td>\n",
       "      <td>1.496 3pg_c + 3.7478 accoa_c + 59.81 atp_c + 0.361 e4p_c + 0.0709 f6p_c + 0.129 g3p_c + 0.205 g6p_c + 0.2557 gln__L_c + 4.9414 glu__L_c + 59.81 h2o_c + 3.547 nad_c + 13.0279 nadph_c + 1.7867 oaa_c + 0.5191 pep_c + 2.8328 pyr_c + 0.8977 r5p_c --&gt; 59.81 adp_c + 4.1182 akg_c + 3.7478 coa_c + 59.81 h_c + 3.547 nadh_c + 13.0279 nadp_c + 59.81 pi_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0.33%</td>\n",
       "      <td>-0.2235</td>\n",
       "      <td>GLNS</td>\n",
       "      <td>atp_c + glu__L_c + nh4_c --&gt; adp_c + gln__L_c + h_c + pi_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10.94%</td>\n",
       "      <td>-7.477</td>\n",
       "      <td>PFK</td>\n",
       "      <td>atp_c + f6p_c --&gt; adp_c + fdp_c + h_c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.metabolite_summary.MetaboliteSummary at 0x1294688fd60>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.atp_c.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>atp_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>ATP C10H12N5O13P3</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01294543e1c0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C10H12N5O13P3</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 13 reaction(s)</strong></td><td>\n",
       "                    GLNS, ATPM, ACKr, PYK, GLNabc, ADK1, BIOMASS_Ecoli_core_w_GAM, PPS, PPCK, PFK, PGK, SUCOAS, ATPS4r</td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite atp_c at 0x1294543e1c0>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.atp_c"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that not all of the thirteen reactions associated with atp_c are carrying flux in this optimization. \n",
    "\n",
    "The shadow prices for each metabolite can be determined through the following command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.866666666666667"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.nadh_c.shadow_price"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "LIsting all the metabolite shadow prices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "glc__D_e    -9.166475e-02\n",
       "gln__L_c    -7.511417e-02\n",
       "gln__L_e    -7.002168e-02\n",
       "glu__L_c    -7.002168e-02\n",
       "glu__L_e    -6.874856e-02\n",
       "glx_c       -1.909682e-02\n",
       "h2o_c        1.084202e-19\n",
       "h2o_e        1.084202e-19\n",
       "h_c          1.273121e-03\n",
       "h_e          4.336809e-19\n",
       "icit_c      -7.129480e-02\n",
       "lac__D_c    -4.201301e-02\n",
       "lac__D_e    -4.073989e-02\n",
       "mal__L_c    -4.837862e-02\n",
       "mal__L_e    -4.583237e-02\n",
       "nad_c        7.638729e-03\n",
       "nadh_c       0.000000e+00\n",
       "nadp_c       8.911850e-03\n",
       "nadph_c      0.000000e+00\n",
       "nh4_c       -0.000000e+00\n",
       "13dpg_c     -4.710549e-02\n",
       "nh4_e       -1.387779e-17\n",
       "o2_c        -0.000000e+00\n",
       "2pg_c       -4.201301e-02\n",
       "o2_e        -0.000000e+00\n",
       "3pg_c       -4.201301e-02\n",
       "oaa_c       -4.201301e-02\n",
       "pep_c       -4.201301e-02\n",
       "6pgc_c      -9.166475e-02\n",
       "pi_c        -1.273121e-03\n",
       "6pgl_c      -9.039162e-02\n",
       "pi_e        -4.163336e-17\n",
       "ac_c        -2.418931e-02\n",
       "pyr_c       -3.564740e-02\n",
       "pyr_e       -3.437428e-02\n",
       "q8_c         2.546243e-03\n",
       "q8h2_c       0.000000e+00\n",
       "r5p_c       -8.275290e-02\n",
       "ru5p__D_c   -8.275290e-02\n",
       "ac_e        -2.291619e-02\n",
       "acald_c     -3.437428e-02\n",
       "s7p_c       -1.133078e-01\n",
       "acald_e     -3.437428e-02\n",
       "accoa_c     -2.800867e-02\n",
       "succ_c      -5.092486e-02\n",
       "succ_e      -4.837862e-02\n",
       "succoa_c    -5.474422e-02\n",
       "acon_C_c    -7.129480e-02\n",
       "xu5p__D_c   -8.275290e-02\n",
       "actp_c      -2.928179e-02\n",
       "adp_c        5.092486e-03\n",
       "akg_c       -6.238295e-02\n",
       "akg_e       -6.110983e-02\n",
       "amp_c        1.018497e-02\n",
       "atp_c        0.000000e+00\n",
       "cit_c       -7.129480e-02\n",
       "co2_c       -0.000000e+00\n",
       "co2_e       -0.000000e+00\n",
       "coa_c        0.000000e+00\n",
       "dhap_c      -5.219798e-02\n",
       "e4p_c       -6.747544e-02\n",
       "etoh_c      -4.073989e-02\n",
       "etoh_e      -3.946677e-02\n",
       "f6p_c       -9.803035e-02\n",
       "fdp_c       -1.043960e-01\n",
       "for_c       -7.638729e-03\n",
       "for_e       -7.638729e-03\n",
       "fru_e       -9.166475e-02\n",
       "fum_c       -4.837862e-02\n",
       "fum_e       -4.583237e-02\n",
       "g3p_c       -5.219798e-02\n",
       "g6p_c       -9.803035e-02\n",
       "Name: shadow_prices, dtype: float64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solution.shadow_prices"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating a table that includes the shadow prices of all the metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metabolite ID</th>\n",
       "      <th>Metabolite Name</th>\n",
       "      <th>Metabolite Formula</th>\n",
       "      <th>Metabolite Compartment</th>\n",
       "      <th>Metabolite Charge</th>\n",
       "      <th>Metabolite Shadow Price</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>D-Glucose</td>\n",
       "      <td>C6H12O6</td>\n",
       "      <td>e</td>\n",
       "      <td>0</td>\n",
       "      <td>5.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gln__L_c</td>\n",
       "      <td>L-Glutamine</td>\n",
       "      <td>C5H10N2O3</td>\n",
       "      <td>c</td>\n",
       "      <td>0</td>\n",
       "      <td>34.155556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gln__L_e</td>\n",
       "      <td>L-Glutamine</td>\n",
       "      <td>C5H10N2O3</td>\n",
       "      <td>e</td>\n",
       "      <td>0</td>\n",
       "      <td>26.911111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>glu__L_c</td>\n",
       "      <td>L-Glutamate</td>\n",
       "      <td>C5H8NO4</td>\n",
       "      <td>c</td>\n",
       "      <td>-1</td>\n",
       "      <td>22.911111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>glu__L_e</td>\n",
       "      <td>L-Glutamate</td>\n",
       "      <td>C5H8NO4</td>\n",
       "      <td>e</td>\n",
       "      <td>-1</td>\n",
       "      <td>22.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>glx_c</td>\n",
       "      <td>Glyoxylate</td>\n",
       "      <td>C2H1O3</td>\n",
       "      <td>c</td>\n",
       "      <td>-1</td>\n",
       "      <td>8.377778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>h2o_c</td>\n",
       "      <td>H2O H2O</td>\n",
       "      <td>H2O</td>\n",
       "      <td>c</td>\n",
       "      <td>0</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>h2o_e</td>\n",
       "      <td>H2O H2O</td>\n",
       "      <td>H2O</td>\n",
       "      <td>e</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>h_c</td>\n",
       "      <td>H+</td>\n",
       "      <td>H</td>\n",
       "      <td>c</td>\n",
       "      <td>1</td>\n",
       "      <td>-2.311111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>h_e</td>\n",
       "      <td>H+</td>\n",
       "      <td>H</td>\n",
       "      <td>e</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>icit_c</td>\n",
       "      <td>Isocitrate</td>\n",
       "      <td>C6H5O7</td>\n",
       "      <td>c</td>\n",
       "      <td>-3</td>\n",
       "      <td>17.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>lac__D_c</td>\n",
       "      <td>D-Lactate</td>\n",
       "      <td>C3H5O3</td>\n",
       "      <td>c</td>\n",
       "      <td>-1</td>\n",
       "      <td>4.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>lac__D_e</td>\n",
       "      <td>D-Lactate</td>\n",
       "      <td>C3H5O3</td>\n",
       "      <td>e</td>\n",
       "      <td>-1</td>\n",
       "      <td>2.355556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>mal__L_c</td>\n",
       "      <td>L-Malate</td>\n",
       "      <td>C4H4O5</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>11.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>mal__L_e</td>\n",
       "      <td>L-Malate</td>\n",
       "      <td>C4H4O5</td>\n",
       "      <td>e</td>\n",
       "      <td>-2</td>\n",
       "      <td>7.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>nad_c</td>\n",
       "      <td>Nicotinamide adenine dinucleotide</td>\n",
       "      <td>C21H26N7O14P2</td>\n",
       "      <td>c</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>nadh_c</td>\n",
       "      <td>Nicotinamide adenine dinucleotide - reduced</td>\n",
       "      <td>C21H27N7O14P2</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>3.866667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>nadp_c</td>\n",
       "      <td>Nicotinamide adenine dinucleotide phosphate</td>\n",
       "      <td>C21H25N7O17P3</td>\n",
       "      <td>c</td>\n",
       "      <td>-3</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>nadph_c</td>\n",
       "      <td>Nicotinamide adenine dinucleotide phosphate - ...</td>\n",
       "      <td>C21H26N7O17P3</td>\n",
       "      <td>c</td>\n",
       "      <td>-4</td>\n",
       "      <td>4.577778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>nh4_c</td>\n",
       "      <td>Ammonium</td>\n",
       "      <td>H4N</td>\n",
       "      <td>c</td>\n",
       "      <td>1</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>13dpg_c</td>\n",
       "      <td>3-Phospho-D-glyceroyl phosphate</td>\n",
       "      <td>C3H4O10P2</td>\n",
       "      <td>c</td>\n",
       "      <td>-4</td>\n",
       "      <td>18.911111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>nh4_e</td>\n",
       "      <td>Ammonium</td>\n",
       "      <td>H4N</td>\n",
       "      <td>e</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>o2_c</td>\n",
       "      <td>O2 O2</td>\n",
       "      <td>O2</td>\n",
       "      <td>c</td>\n",
       "      <td>0</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2pg_c</td>\n",
       "      <td>D-Glycerate 2-phosphate</td>\n",
       "      <td>C3H4O7P</td>\n",
       "      <td>c</td>\n",
       "      <td>-3</td>\n",
       "      <td>11.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>o2_e</td>\n",
       "      <td>O2 O2</td>\n",
       "      <td>O2</td>\n",
       "      <td>e</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>3pg_c</td>\n",
       "      <td>3-Phospho-D-glycerate</td>\n",
       "      <td>C3H4O7P</td>\n",
       "      <td>c</td>\n",
       "      <td>-3</td>\n",
       "      <td>10.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>oaa_c</td>\n",
       "      <td>Oxaloacetate</td>\n",
       "      <td>C4H2O5</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>10.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>pep_c</td>\n",
       "      <td>Phosphoenolpyruvate</td>\n",
       "      <td>C3H2O6P</td>\n",
       "      <td>c</td>\n",
       "      <td>-3</td>\n",
       "      <td>14.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>6pgc_c</td>\n",
       "      <td>6-Phospho-D-gluconate</td>\n",
       "      <td>C6H10O10P</td>\n",
       "      <td>c</td>\n",
       "      <td>-3</td>\n",
       "      <td>16.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>pi_c</td>\n",
       "      <td>Phosphate</td>\n",
       "      <td>HO4P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>3.311111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>6pgl_c</td>\n",
       "      <td>6-phospho-D-glucono-1,5-lactone</td>\n",
       "      <td>C6H9O9P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>15.488889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>pi_e</td>\n",
       "      <td>Phosphate</td>\n",
       "      <td>HO4P</td>\n",
       "      <td>e</td>\n",
       "      <td>-2</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ac_c</td>\n",
       "      <td>Acetate</td>\n",
       "      <td>C2H3O2</td>\n",
       "      <td>c</td>\n",
       "      <td>-1</td>\n",
       "      <td>-0.688889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>pyr_c</td>\n",
       "      <td>Pyruvate</td>\n",
       "      <td>C3H3O3</td>\n",
       "      <td>c</td>\n",
       "      <td>-1</td>\n",
       "      <td>4.111111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>pyr_e</td>\n",
       "      <td>Pyruvate</td>\n",
       "      <td>C3H3O3</td>\n",
       "      <td>e</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>q8_c</td>\n",
       "      <td>Ubiquinone-8</td>\n",
       "      <td>C49H74O4</td>\n",
       "      <td>c</td>\n",
       "      <td>0</td>\n",
       "      <td>1.377778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>q8h2_c</td>\n",
       "      <td>Ubiquinol-8</td>\n",
       "      <td>C49H76O4</td>\n",
       "      <td>c</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>r5p_c</td>\n",
       "      <td>Alpha-D-Ribose 5-phosphate</td>\n",
       "      <td>C5H9O8P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>16.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ru5p__D_c</td>\n",
       "      <td>D-Ribulose 5-phosphate</td>\n",
       "      <td>C5H9O8P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>15.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>ac_e</td>\n",
       "      <td>Acetate</td>\n",
       "      <td>C2H3O2</td>\n",
       "      <td>e</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>acald_c</td>\n",
       "      <td>Acetaldehyde</td>\n",
       "      <td>C2H4O</td>\n",
       "      <td>c</td>\n",
       "      <td>0</td>\n",
       "      <td>3.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>s7p_c</td>\n",
       "      <td>Sedoheptulose 7-phosphate</td>\n",
       "      <td>C7H13O10P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>17.288889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>acald_e</td>\n",
       "      <td>Acetaldehyde</td>\n",
       "      <td>C2H4O</td>\n",
       "      <td>e</td>\n",
       "      <td>0</td>\n",
       "      <td>2.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>accoa_c</td>\n",
       "      <td>Acetyl-CoA</td>\n",
       "      <td>C23H34N7O17P3S</td>\n",
       "      <td>c</td>\n",
       "      <td>-4</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>succ_c</td>\n",
       "      <td>Succinate</td>\n",
       "      <td>C4H4O4</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>9.844444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>succ_e</td>\n",
       "      <td>Succinate</td>\n",
       "      <td>C4H4O4</td>\n",
       "      <td>e</td>\n",
       "      <td>-2</td>\n",
       "      <td>6.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>succoa_c</td>\n",
       "      <td>Succinyl-CoA</td>\n",
       "      <td>C25H35N7O19P3S</td>\n",
       "      <td>c</td>\n",
       "      <td>-5</td>\n",
       "      <td>11.533333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>acon_C_c</td>\n",
       "      <td>Cis-Aconitate</td>\n",
       "      <td>C6H3O6</td>\n",
       "      <td>c</td>\n",
       "      <td>-3</td>\n",
       "      <td>18.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>xu5p__D_c</td>\n",
       "      <td>D-Xylulose 5-phosphate</td>\n",
       "      <td>C5H9O8P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>16.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>actp_c</td>\n",
       "      <td>Acetyl phosphate</td>\n",
       "      <td>C2H3O5P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>7.555556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>adp_c</td>\n",
       "      <td>ADP C10H12N5O10P2</td>\n",
       "      <td>C10H12N5O10P2</td>\n",
       "      <td>c</td>\n",
       "      <td>-3</td>\n",
       "      <td>-9.244444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>akg_c</td>\n",
       "      <td>2-Oxoglutarate</td>\n",
       "      <td>C5H4O5</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>15.644444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>akg_e</td>\n",
       "      <td>2-Oxoglutarate</td>\n",
       "      <td>C5H4O5</td>\n",
       "      <td>e</td>\n",
       "      <td>-2</td>\n",
       "      <td>13.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>amp_c</td>\n",
       "      <td>AMP C10H12N5O7P</td>\n",
       "      <td>C10H12N5O7P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>-17.488889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>atp_c</td>\n",
       "      <td>ATP C10H12N5O13P3</td>\n",
       "      <td>C10H12N5O13P3</td>\n",
       "      <td>c</td>\n",
       "      <td>-4</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>cit_c</td>\n",
       "      <td>Citrate</td>\n",
       "      <td>C6H5O7</td>\n",
       "      <td>c</td>\n",
       "      <td>-3</td>\n",
       "      <td>15.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>co2_c</td>\n",
       "      <td>CO2 CO2</td>\n",
       "      <td>CO2</td>\n",
       "      <td>c</td>\n",
       "      <td>0</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>co2_e</td>\n",
       "      <td>CO2 CO2</td>\n",
       "      <td>CO2</td>\n",
       "      <td>e</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>coa_c</td>\n",
       "      <td>Coenzyme A</td>\n",
       "      <td>C21H32N7O16P3S</td>\n",
       "      <td>c</td>\n",
       "      <td>-4</td>\n",
       "      <td>-3.244444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>dhap_c</td>\n",
       "      <td>Dihydroxyacetone phosphate</td>\n",
       "      <td>C3H5O6P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>15.155556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>e4p_c</td>\n",
       "      <td>D-Erythrose 4-phosphate</td>\n",
       "      <td>C4H7O7P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>16.688889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>etoh_c</td>\n",
       "      <td>Ethanol</td>\n",
       "      <td>C2H6O</td>\n",
       "      <td>c</td>\n",
       "      <td>0</td>\n",
       "      <td>4.355556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>etoh_e</td>\n",
       "      <td>Ethanol</td>\n",
       "      <td>C2H6O</td>\n",
       "      <td>e</td>\n",
       "      <td>0</td>\n",
       "      <td>2.044444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>f6p_c</td>\n",
       "      <td>D-Fructose 6-phosphate</td>\n",
       "      <td>C6H11O9P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>17.755556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>fdp_c</td>\n",
       "      <td>D-Fructose 1,6-bisphosphate</td>\n",
       "      <td>C6H10O12P2</td>\n",
       "      <td>c</td>\n",
       "      <td>-4</td>\n",
       "      <td>30.311111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>for_c</td>\n",
       "      <td>Formate</td>\n",
       "      <td>CH1O2</td>\n",
       "      <td>c</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>for_e</td>\n",
       "      <td>Formate</td>\n",
       "      <td>CH1O2</td>\n",
       "      <td>e</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>fru_e</td>\n",
       "      <td>D-Fructose</td>\n",
       "      <td>C6H12O6</td>\n",
       "      <td>e</td>\n",
       "      <td>0</td>\n",
       "      <td>6.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>fum_c</td>\n",
       "      <td>Fumarate</td>\n",
       "      <td>C4H2O4</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>12.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>fum_e</td>\n",
       "      <td>Fumarate</td>\n",
       "      <td>C4H2O4</td>\n",
       "      <td>e</td>\n",
       "      <td>-2</td>\n",
       "      <td>8.600000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>g3p_c</td>\n",
       "      <td>Glyceraldehyde 3-phosphate</td>\n",
       "      <td>C3H5O6P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>16.155556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>g6p_c</td>\n",
       "      <td>D-Glucose 6-phosphate</td>\n",
       "      <td>C6H11O9P</td>\n",
       "      <td>c</td>\n",
       "      <td>-2</td>\n",
       "      <td>16.755556</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Metabolite ID                                    Metabolite Name  \\\n",
       "0       glc__D_e                                          D-Glucose   \n",
       "1       gln__L_c                                        L-Glutamine   \n",
       "2       gln__L_e                                        L-Glutamine   \n",
       "3       glu__L_c                                        L-Glutamate   \n",
       "4       glu__L_e                                        L-Glutamate   \n",
       "5          glx_c                                         Glyoxylate   \n",
       "6          h2o_c                                            H2O H2O   \n",
       "7          h2o_e                                            H2O H2O   \n",
       "8            h_c                                                 H+   \n",
       "9            h_e                                                 H+   \n",
       "10        icit_c                                         Isocitrate   \n",
       "11      lac__D_c                                          D-Lactate   \n",
       "12      lac__D_e                                          D-Lactate   \n",
       "13      mal__L_c                                           L-Malate   \n",
       "14      mal__L_e                                           L-Malate   \n",
       "15         nad_c                  Nicotinamide adenine dinucleotide   \n",
       "16        nadh_c        Nicotinamide adenine dinucleotide - reduced   \n",
       "17        nadp_c        Nicotinamide adenine dinucleotide phosphate   \n",
       "18       nadph_c  Nicotinamide adenine dinucleotide phosphate - ...   \n",
       "19         nh4_c                                           Ammonium   \n",
       "20       13dpg_c                    3-Phospho-D-glyceroyl phosphate   \n",
       "21         nh4_e                                           Ammonium   \n",
       "22          o2_c                                              O2 O2   \n",
       "23         2pg_c                            D-Glycerate 2-phosphate   \n",
       "24          o2_e                                              O2 O2   \n",
       "25         3pg_c                              3-Phospho-D-glycerate   \n",
       "26         oaa_c                                       Oxaloacetate   \n",
       "27         pep_c                                Phosphoenolpyruvate   \n",
       "28        6pgc_c                              6-Phospho-D-gluconate   \n",
       "29          pi_c                                          Phosphate   \n",
       "30        6pgl_c                    6-phospho-D-glucono-1,5-lactone   \n",
       "31          pi_e                                          Phosphate   \n",
       "32          ac_c                                            Acetate   \n",
       "33         pyr_c                                           Pyruvate   \n",
       "34         pyr_e                                           Pyruvate   \n",
       "35          q8_c                                       Ubiquinone-8   \n",
       "36        q8h2_c                                        Ubiquinol-8   \n",
       "37         r5p_c                         Alpha-D-Ribose 5-phosphate   \n",
       "38     ru5p__D_c                             D-Ribulose 5-phosphate   \n",
       "39          ac_e                                            Acetate   \n",
       "40       acald_c                                       Acetaldehyde   \n",
       "41         s7p_c                          Sedoheptulose 7-phosphate   \n",
       "42       acald_e                                       Acetaldehyde   \n",
       "43       accoa_c                                         Acetyl-CoA   \n",
       "44        succ_c                                          Succinate   \n",
       "45        succ_e                                          Succinate   \n",
       "46      succoa_c                                       Succinyl-CoA   \n",
       "47      acon_C_c                                      Cis-Aconitate   \n",
       "48     xu5p__D_c                             D-Xylulose 5-phosphate   \n",
       "49        actp_c                                   Acetyl phosphate   \n",
       "50         adp_c                                  ADP C10H12N5O10P2   \n",
       "51         akg_c                                     2-Oxoglutarate   \n",
       "52         akg_e                                     2-Oxoglutarate   \n",
       "53         amp_c                                    AMP C10H12N5O7P   \n",
       "54         atp_c                                  ATP C10H12N5O13P3   \n",
       "55         cit_c                                            Citrate   \n",
       "56         co2_c                                            CO2 CO2   \n",
       "57         co2_e                                            CO2 CO2   \n",
       "58         coa_c                                         Coenzyme A   \n",
       "59        dhap_c                         Dihydroxyacetone phosphate   \n",
       "60         e4p_c                            D-Erythrose 4-phosphate   \n",
       "61        etoh_c                                            Ethanol   \n",
       "62        etoh_e                                            Ethanol   \n",
       "63         f6p_c                             D-Fructose 6-phosphate   \n",
       "64         fdp_c                        D-Fructose 1,6-bisphosphate   \n",
       "65         for_c                                            Formate   \n",
       "66         for_e                                            Formate   \n",
       "67         fru_e                                         D-Fructose   \n",
       "68         fum_c                                           Fumarate   \n",
       "69         fum_e                                           Fumarate   \n",
       "70         g3p_c                         Glyceraldehyde 3-phosphate   \n",
       "71         g6p_c                              D-Glucose 6-phosphate   \n",
       "\n",
       "   Metabolite Formula Metabolite Compartment  Metabolite Charge  \\\n",
       "0             C6H12O6                      e                  0   \n",
       "1           C5H10N2O3                      c                  0   \n",
       "2           C5H10N2O3                      e                  0   \n",
       "3             C5H8NO4                      c                 -1   \n",
       "4             C5H8NO4                      e                 -1   \n",
       "5              C2H1O3                      c                 -1   \n",
       "6                 H2O                      c                  0   \n",
       "7                 H2O                      e                  0   \n",
       "8                   H                      c                  1   \n",
       "9                   H                      e                  1   \n",
       "10             C6H5O7                      c                 -3   \n",
       "11             C3H5O3                      c                 -1   \n",
       "12             C3H5O3                      e                 -1   \n",
       "13             C4H4O5                      c                 -2   \n",
       "14             C4H4O5                      e                 -2   \n",
       "15      C21H26N7O14P2                      c                 -1   \n",
       "16      C21H27N7O14P2                      c                 -2   \n",
       "17      C21H25N7O17P3                      c                 -3   \n",
       "18      C21H26N7O17P3                      c                 -4   \n",
       "19                H4N                      c                  1   \n",
       "20          C3H4O10P2                      c                 -4   \n",
       "21                H4N                      e                  1   \n",
       "22                 O2                      c                  0   \n",
       "23            C3H4O7P                      c                 -3   \n",
       "24                 O2                      e                  0   \n",
       "25            C3H4O7P                      c                 -3   \n",
       "26             C4H2O5                      c                 -2   \n",
       "27            C3H2O6P                      c                 -3   \n",
       "28          C6H10O10P                      c                 -3   \n",
       "29               HO4P                      c                 -2   \n",
       "30            C6H9O9P                      c                 -2   \n",
       "31               HO4P                      e                 -2   \n",
       "32             C2H3O2                      c                 -1   \n",
       "33             C3H3O3                      c                 -1   \n",
       "34             C3H3O3                      e                 -1   \n",
       "35           C49H74O4                      c                  0   \n",
       "36           C49H76O4                      c                  0   \n",
       "37            C5H9O8P                      c                 -2   \n",
       "38            C5H9O8P                      c                 -2   \n",
       "39             C2H3O2                      e                 -1   \n",
       "40              C2H4O                      c                  0   \n",
       "41          C7H13O10P                      c                 -2   \n",
       "42              C2H4O                      e                  0   \n",
       "43     C23H34N7O17P3S                      c                 -4   \n",
       "44             C4H4O4                      c                 -2   \n",
       "45             C4H4O4                      e                 -2   \n",
       "46     C25H35N7O19P3S                      c                 -5   \n",
       "47             C6H3O6                      c                 -3   \n",
       "48            C5H9O8P                      c                 -2   \n",
       "49            C2H3O5P                      c                 -2   \n",
       "50      C10H12N5O10P2                      c                 -3   \n",
       "51             C5H4O5                      c                 -2   \n",
       "52             C5H4O5                      e                 -2   \n",
       "53        C10H12N5O7P                      c                 -2   \n",
       "54      C10H12N5O13P3                      c                 -4   \n",
       "55             C6H5O7                      c                 -3   \n",
       "56                CO2                      c                  0   \n",
       "57                CO2                      e                  0   \n",
       "58     C21H32N7O16P3S                      c                 -4   \n",
       "59            C3H5O6P                      c                 -2   \n",
       "60            C4H7O7P                      c                 -2   \n",
       "61              C2H6O                      c                  0   \n",
       "62              C2H6O                      e                  0   \n",
       "63           C6H11O9P                      c                 -2   \n",
       "64         C6H10O12P2                      c                 -4   \n",
       "65              CH1O2                      c                 -1   \n",
       "66              CH1O2                      e                 -1   \n",
       "67            C6H12O6                      e                  0   \n",
       "68             C4H2O4                      c                 -2   \n",
       "69             C4H2O4                      e                 -2   \n",
       "70            C3H5O6P                      c                 -2   \n",
       "71           C6H11O9P                      c                 -2   \n",
       "\n",
       "    Metabolite Shadow Price  \n",
       "0                  5.200000  \n",
       "1                 34.155556  \n",
       "2                 26.911111  \n",
       "3                 22.911111  \n",
       "4                 22.600000  \n",
       "5                  8.377778  \n",
       "6                 -2.000000  \n",
       "7                 -1.000000  \n",
       "8                 -2.311111  \n",
       "9                 -1.000000  \n",
       "10                17.222222  \n",
       "11                 4.666667  \n",
       "12                 2.355556  \n",
       "13                11.222222  \n",
       "14                 7.600000  \n",
       "15                 0.000000  \n",
       "16                 3.866667  \n",
       "17                 0.000000  \n",
       "18                 4.577778  \n",
       "19                 2.000000  \n",
       "20                18.911111  \n",
       "21                 1.000000  \n",
       "22                 2.000000  \n",
       "23                11.666667  \n",
       "24                 1.000000  \n",
       "25                10.666667  \n",
       "26                10.666667  \n",
       "27                14.666667  \n",
       "28                16.800000  \n",
       "29                 3.311111  \n",
       "30                15.488889  \n",
       "31                 1.000000  \n",
       "32                -0.688889  \n",
       "33                 4.111111  \n",
       "34                 1.800000  \n",
       "35                 1.377778  \n",
       "36                 0.000000  \n",
       "37                16.222222  \n",
       "38                15.222222  \n",
       "39                -1.000000  \n",
       "40                 3.800000  \n",
       "41                17.288889  \n",
       "42                 2.800000  \n",
       "43                 0.000000  \n",
       "44                 9.844444  \n",
       "45                 6.222222  \n",
       "46                11.533333  \n",
       "47                18.222222  \n",
       "48                16.222222  \n",
       "49                 7.555556  \n",
       "50                -9.244444  \n",
       "51                15.644444  \n",
       "52                13.333333  \n",
       "53               -17.488889  \n",
       "54                 0.000000  \n",
       "55                15.222222  \n",
       "56                -2.000000  \n",
       "57                -1.000000  \n",
       "58                -3.244444  \n",
       "59                15.155556  \n",
       "60                16.688889  \n",
       "61                 4.355556  \n",
       "62                 2.044444  \n",
       "63                17.755556  \n",
       "64                30.311111  \n",
       "65                -2.000000  \n",
       "66                -1.000000  \n",
       "67                 6.200000  \n",
       "68                12.222222  \n",
       "69                 8.600000  \n",
       "70                16.155556  \n",
       "71                16.755556  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolite_names = [m.name for m in model.metabolites]\n",
    "metabolite_ids = [m.id for m in model.metabolites]\n",
    "metabolite_compartment = [m.compartment for m in model.metabolites]\n",
    "metabolite_formula = [m.formula for m in model.metabolites]\n",
    "metabolite_charge = [m.charge for m in model.metabolites]\n",
    "metabolite_SP = [m.shadow_price for m in model.metabolites]\n",
    "metaboliteList = {'Metabolite ID': metabolite_ids,\n",
    "                  'Metabolite Name': metabolite_names,\n",
    "                  'Metabolite Formula': metabolite_formula,\n",
    "                  'Metabolite Compartment': metabolite_compartment,\n",
    "                  'Metabolite Charge': metabolite_charge,\n",
    "                  'Metabolite Shadow Price': metabolite_SP\n",
    "                 }\n",
    "\n",
    "df = pd.DataFrame(metaboliteList, columns= ['Metabolite ID','Metabolite Name', 'Metabolite Formula', 'Metabolite Compartment','Metabolite Charge','Metabolite Shadow Price'])\n",
    "pd.set_option('display.max_rows', 500)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Changing the Objective Function\n",
    "The objective function is determined from the objective_coefficient attribute of the objective reaction(s). For most cases, a biomass function which describes the composition of metabolites which make up a cell is used. The objective function can be change to a single or multiple reactions.\n",
    "\n",
    "The objective function for a given model can be determined using the **model.summary()** or the **linear_reaction_coefficients(model)** commands."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 BIOMASS_Ecoli_core_w_GAM = 0.8739215069684312</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>glc__D_e</td>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>100.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e</td>\n",
       "      <td>4.765</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e</td>\n",
       "      <td>21.8</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e</td>\n",
       "      <td>3.215</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e</td>\n",
       "      <td>-22.81</td>\n",
       "      <td>1</td>\n",
       "      <td>100.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e</td>\n",
       "      <td>-29.18</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e</td>\n",
       "      <td>-17.53</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x129468f2040>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "or"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{<Reaction BIOMASS_Ecoli_core_w_GAM at 0x1294548f6d0>: 1.0}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from cobra.util.solver import linear_reaction_coefficients\n",
    "linear_reaction_coefficients(model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The details of the objective function can be found by using the following code(the default objective function for the *E.coli* core model is the biomass function)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>BIOMASS_Ecoli_core_w_GAM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Biomass Objective Function with GAM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01294548f6d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>1.496 3pg_c + 3.7478 accoa_c + 59.81 atp_c + 0.361 e4p_c + 0.0709 f6p_c + 0.129 g3p_c + 0.205 g6p_c + 0.2557 gln__L_c + 4.9414 glu__L_c + 59.81 h2o_c + 3.547 nad_c + 13.0279 nadph_c + 1.7867 oaa_c...</p>\n",
       "                    <p style='text-align:right'>1.496 3-Phospho-D-glycerate + 3.7478 Acetyl-CoA + 59.81 ATP C10H12N5O13P3 + 0.361 D-Erythrose 4-phosphate + 0.0709 D-Fructose 6-phosphate + 0.129 Glyceraldehyde 3-phosphate + 0.205 D-Glucose...</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction BIOMASS_Ecoli_core_w_GAM at 0x1294548f6d0>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biomass_rxn = model.reactions.get_by_id(\"BIOMASS_Ecoli_core_w_GAM\")\n",
    "biomass_rxn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Currently in the model, there is only one reaction in the objective (the biomass reaction), with an linear coefficient of 1."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The objective function can be changed by assigning model.objective, which can be a reaction object (or just its name), or a dict of {Reaction: objective_coefficient}."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# change the objective to ATPM\n",
    "model.objective = \"ATPM\"\n",
    "\n",
    "# The upper bound should be 1000, so that we get\n",
    "# the actual optimal value\n",
    "model.reactions.get_by_id(\"ATPM\").upper_bound = 1000."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Verifying this change to the objective function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{<Reaction ATPM at 0x12945471430>: 1.0}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linear_reaction_coefficients(model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The characteristics of the new objective function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ATPM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>ATP maintenance requirement</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x012945471430</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>atp_c + h2o_c --> adp_c + h_c + pi_c</p>\n",
       "                    <p style='text-align:right'>ATP C10H12N5O13P3 + H2O H2O --> ADP C10H12N5O10P2 + H+ + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>8.39</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ATPM at 0x12945471430>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "objectiveFunction = model.reactions.get_by_id(\"ATPM\")\n",
    "objectiveFunction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The numerical value of the new ATPM objective function is given below. Notice that the model is no longer optimizing for cell growth but now it is maximizing the production of ATP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "175.0000000000001"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.optimize().objective_value"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running Geometric FBA\n",
    "Geometric FBA finds a unique optimal flux distribution which is central to the range of possible fluxes. For more details on geometric FBA, please see K Smallbone, E Simeonidis (2009)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 0.000</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PFK</th>\n",
       "      <td>7.477382</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PFL</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGI</th>\n",
       "      <td>4.860861</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGK</th>\n",
       "      <td>-16.023526</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGL</th>\n",
       "      <td>4.959985</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NADH16</th>\n",
       "      <td>38.534610</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NADTRHD</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH4t</th>\n",
       "      <td>4.765319</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2t</th>\n",
       "      <td>21.799493</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDH</th>\n",
       "      <td>9.282533</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>95 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 0.000 at 0x12946818880>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.objective = \"BIOMASS_Ecoli_core_w_GAM\"\n",
    "geometric_fba_sol = cobra.flux_analysis.geometric_fba(model)\n",
    "geometric_fba_sol"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting the Fluxes on an Escher Map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import escher\n",
    "from escher import Builder"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To plot the maps of the different models requires a Jupyter widget called \"Builder.\" This widget will execute in a Jupyter cell to embed an Escher map. Two names are required for the widget to operate properly; 1) the \"map_name\" is the name of the map file (different from the model file) that contains all the information to plot a map of the organism, 2) the \"model_name\" is the name of a genome-scale metabolic network reconstruction or COBRA model. The maps and models that are supported by the Escher website include the following."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'organism': 'Saccharomyces cerevisiae',\n",
       "  'map_name': 'iMM904.Central carbon metabolism'},\n",
       " {'organism': 'Homo sapiens',\n",
       "  'map_name': 'RECON1.Inositol retinol metabolism'},\n",
       " {'organism': 'Homo sapiens', 'map_name': 'RECON1.Glycolysis TCA PPP'},\n",
       " {'organism': 'Homo sapiens', 'map_name': 'RECON1.Tryptophan metabolism'},\n",
       " {'organism': 'Homo sapiens', 'map_name': 'RECON1.Carbohydrate metabolism'},\n",
       " {'organism': 'Homo sapiens',\n",
       "  'map_name': 'RECON1.Amino acid metabolism (partial)'},\n",
       " {'organism': 'Escherichia coli', 'map_name': 'iJO1366.Nucleotide metabolism'},\n",
       " {'organism': 'Escherichia coli',\n",
       "  'map_name': 'iJO1366.Fatty acid biosynthesis (saturated)'},\n",
       " {'organism': 'Escherichia coli',\n",
       "  'map_name': 'iJO1366.Nucleotide and histidine biosynthesis'},\n",
       " {'organism': 'Escherichia coli', 'map_name': 'e_coli_core.Core metabolism'},\n",
       " {'organism': 'Escherichia coli', 'map_name': 'iJO1366.Central metabolism'},\n",
       " {'organism': 'Escherichia coli',\n",
       "  'map_name': 'iJO1366.Fatty acid beta-oxidation'}]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "escher.list_available_maps()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'organism': 'Saccharomyces cerevisiae', 'model_name': 'iMM904'},\n",
       " {'organism': 'Homo sapiens', 'model_name': 'RECON1'},\n",
       " {'organism': 'Escherichia coli', 'model_name': 'e_coli_core'},\n",
       " {'organism': 'Escherichia coli', 'model_name': 'iJO1366'}]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "escher.list_available_models()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To load a map and model into the Builder widget can be done as follows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Map from https://escher.github.io/1-0-0/6/maps/Escherichia%20coli/e_coli_core.Core%20metabolism.json\n",
      "Downloading Model from https://escher.github.io/1-0-0/6/models/Escherichia%20coli/e_coli_core.json\n"
     ]
    }
   ],
   "source": [
    "builder = Builder(\n",
    "    map_name='e_coli_core.Core metabolism', \n",
    "    model_name='e_coli_core', # This loads the model name into Escher but does not change the model in the COBRApy simulation\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sets the model and map name for the Escher package. The model name that is downloaded does not replace the model used in the COBRApy simulations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bc88e07fcb1d4aefbf101338a453d4a6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Builder()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "builder"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that the flux data is overlayed on the map that was created by the Builder widget. Plotting the flux data from an FBA simulation (model.optimize()) onto the map. If you zoom in on the map and click the arrow to the left you can put the arrow over a reaction or metabolite and see the flux value(reactions). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "solution = model.optimize()\n",
    "builder.reaction_data = solution.fluxes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now lets create a new map that also includes the shadow prices of the metabolites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Map from https://escher.github.io/1-0-0/6/maps/Escherichia%20coli/e_coli_core.Core%20metabolism.json\n",
      "Downloading Model from https://escher.github.io/1-0-0/6/models/Escherichia%20coli/e_coli_core.json\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "63978c44db45424da501734aafd67e09",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Builder()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "builder = Builder(\n",
    "    map_name='e_coli_core.Core metabolism', \n",
    "    model_name='e_coli_core', \n",
    ")\n",
    "builder"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you zoom in on the map and click the arrow to the left you can move the arrow over a metabolite's name and the shadow price information for the metabolite will be shown. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "solution = model.optimize()\n",
    "builder.reaction_data = solution.fluxes\n",
    "builder.metabolite_data = solution.shadow_prices"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "More information on how to use Escher can be found at https://escher.readthedocs.io/en/latest/index.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map Animation\n",
    "\n",
    "Animations are relatively easy with an Escher map. Here's a simple example where flux predictions are swept across and range to update the map of *E. coli* core metabolism.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading Map from https://escher.github.io/1-0-0/6/maps/Escherichia%20coli/e_coli_core.Core%20metabolism.json\n",
      "Downloading Model from https://escher.github.io/1-0-0/6/models/Escherichia%20coli/e_coli_core.json\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7c0c2534d2e24851a936ccb1f9f9d616",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Builder(height=600)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "builder = Builder(\n",
    "    height=600, \n",
    "    map_name='e_coli_core.Core metabolism', \n",
    "    model_name='e_coli_core',\n",
    ")\n",
    "builder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "from time import sleep\n",
    "timestep = 0.5\n",
    "lim = [0, 0.5]\n",
    "for lb in range(-30,1,1):\n",
    "    model.reactions.EX_o2_e.lower_bound = lb\n",
    "    solution = model.optimize()\n",
    "    builder.reaction_data = solution.fluxes\n",
    "    sleep(timestep)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this simulation, the o2 level starts at -30 mmol/gDW-hr which puts the organism in an aerobic state. As the o2 level drops to 0 mmol/gDW-hr, the organism passes through an anoxic state on it's way to an anaerobic state"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saving Maps as HTML\n",
    "\n",
    "The Escher Builder can also be saved as a standalone HTML file, which you can view by opening in a browser. Just provide a filepath, and the map will be bundled along with all the current options.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "builder.save_html('example_map.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "1. https://cobrapy.readthedocs.io/en/latest/simulating.html#\n",
    "2. https://escher.readthedocs.io/en/latest/index.html"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
